### DISCUSSION TITLE
What can we do to improve performance?

### DISCUSSION THREAD

**Discussion by manugill** [pixiv/three-vrm#1225] [General]
I love this library, have been attempting to build a game with it but it seems to really only be optimised to the point of rendering a single VRM character using three.js without leaving perf headroom to do much else (unless your computer is extremely powerful with a GPU). We can definitely improve the rendering for mobile devices but currently the library is unusable for many VR usecases as well (like a web based VR chat).

I have been trying to do this/looking into this and found a few things:

- In my testing I found that MMD/PMD models tend to perform far better. So, sadly for my current project, I'm going to have to go with those. But by design, it should be less intensive to render VRM models as the format was designed for realtime rendering ( with its lightweight physics capabilities and other optimisations) while MMD wasn't. I'd be quite excited to see this library become mature enough to exceed the performance of MMD models at some point. Would be happy to contribute to that as well.

- Another thing is that it's not possible to load a single VRM model reuse it - we seem to be only able to render a single .vrm file of a kind then there's bugs around even attempting to do this with breaking poses, etc. This would be very useful rendering mobs in RPGs or something low quality background characters, etc. One usecase I'd be interested in is using a simple low quality VRM model for multiple users in a multiplayer game. Would be keen to contribute to this as well, seems like it shouldn't be that hard?

- Found this research paper where they do a perf comparison with MMD too and find that optimising the GLTF loader resulted in a 4x performance increase: https://www.researchgate.net/publication/366472491_A_Method_to_Optimize_the_Performance_of_Three-VRM. The actual code wasn't shared though.

**Comment by 0b5vr** [comment]
Thank you for the discussion! Performance is a thing we always love to improve working on this library.

- Would you mind if I ask you to provide working examples for MMD vs. VRM? Then we can proceed to profile them and find what is the difference between them. Having a demo on sandbox platforms like Glitch or CodeSandbox will be the best.

- The performance issue when we reuse a single model might be solved by addressing the cloning feature we are discussing on #1172 .

- I suspect that the paper you provided addresses model-specific issues instead of ideas that can apply to general models, and does not provide the model source, as well as the actual code as you mentioned. I think we can't really replicate the study.

**Comment by hotairbag** [comment]
Would also love to see some general performance improvements, especially for mobile web browsers!
